src/buffer.rs

```diff
@@ -320,6 +322,36 @@ pub trait StringDecoder {
     fn decode_string(data: &[u8], cursor: &mut usize, delimiter: Self::Delimiter) -> GDResult<String>;
 }
 
+pub struct Latin1Decoder;
+
+impl StringDecoder for Latin1Decoder {
+    type Delimiter = [u8; 1];
+
+    const DELIMITER: Self::Delimiter = [0x00];
+
+    fn decode_string(data: &[u8], cursor: &mut usize, delimiter: Self::Delimiter) -> GDResult<String> {
```

Comment:
Looking at this we are starting to have repetition on finding the delimiter, It might look cleaner with a helper function to handle this. Something like this would be cleaner I think.

```rs
/// Finds the position of the delimiter in the data slice.
///
/// # Arguments
///
/// * `data` - The data slice to search for the delimiter in.
/// * `delimiter` - The delimiter to search for in the data slice.
///
/// # Returns
///
/// The position of the first occurrence of the delimiter in the data slice.
/// If the delimiter is not found, returns the length of the data slice.
fn find_delimiter_position(data: &[u8], delimiter: &[u8]) -> usize {
    data.windows(delimiter.len())
        .position(|window| window == delimiter)
        .unwrap_or(data.len())
}
```

```rs
pub struct Latin1Decoder;

impl StringDecoder for Latin1Decoder {
    type Delimiter = [u8; 1];

    const DELIMITER: Self::Delimiter = [0x00];

    fn decode_string(data: &[u8], cursor: &mut usize, delimiter: Self::Delimiter) -> GDResult<String> {
        let position = find_delimiter_position(data, &delimiter);
        
        let result = encoding::all::ISO_8859_1
            .decode(&data[..position], DecoderTrap::Strict)
            .map_err(|e| PacketBad.context(e))?;

        *cursor += position + delimiter.len();

        Ok(result)
    }
}
```
```rs
pub struct Utf8Decoder;

impl StringDecoder for Utf8Decoder {
    type Delimiter = [u8; 1];

    const DELIMITER: Self::Delimiter = [0x00];

    fn decode_string(data: &[u8], cursor: &mut usize, delimiter: Self::Delimiter) -> GDResult<String> {
        let position = find_delimiter_position(data, &delimiter);

        let result = std::str::from_utf8(&data[..position])
            .map_err(|e| PacketBad.context(e))?
            .to_owned();

        *cursor += position + delimiter.len();

        Ok(result)
    }
}
```
```rs
pub struct Utf16Decoder<B: ByteOrder> {
    _marker: PhantomData<B>,
}

impl<B: ByteOrder> StringDecoder for Utf16Decoder<B> {
    type Delimiter = [u8; 2];

    const DELIMITER: Self::Delimiter = [0x00, 0x00];

    fn decode_string(data: &[u8], cursor: &mut usize, delimiter: Self::Delimiter) -> GDResult<String> {
        let position = find_delimiter_position(data, &delimiter);

        let mut paired_buf: Vec<u16> = vec![0; position / 2];
        B::read_u16_into(&data[..position], &mut paired_buf);

        let result = String::from_utf16(&paired_buf).map_err(|e| PacketBad.context(e))?;

        *cursor += position + delimiter.len();

        Ok(result)
    }
}
```